core/commands/ping.go

```diff
@@ -59,166 +59,114 @@ trip latency information.
 			return ErrNotOnline
 		}
 
-		addr, peerID, err := ParsePeerParam(req.Arguments[0])
+		addr, pid, err := ParsePeerParam(req.Arguments[0])
 		if err != nil {
 			return fmt.Errorf("failed to parse peer address '%s': %s", req.Arguments[0], err)
 		}
 
-		if peerID == n.Identity {
+		if pid == n.Identity {
 			return ErrPingSelf
 		}
 
 		if addr != nil {
-			n.Peerstore.AddAddr(peerID, addr, pstore.TempAddrTTL) // temporary
+			n.Peerstore.AddAddr(pid, addr, pstore.TempAddrTTL) // temporary
 		}
 
 		numPings, _ := req.Options[pingCountOptionName].(int)
 		if numPings <= 0 {
 			return fmt.Errorf("error: ping count must be greater than 0, was %d", numPings)
 		}
 
-		outChan := pingPeer(req.Context, n, peerID, numPings)
-
-		return res.Emit(outChan)
-	},
-	Type: PingResult{},
-	Encoders: cmds.EncoderMap{
-		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, out *PingResult) error {
-			if len(out.Text) > 0 {
-				fmt.Fprintln(w, out.Text)
-			} else if out.Success {
-				fmt.Fprintf(w, "Pong received: time=%.2f ms\n", out.Time.Seconds()*1000)
-			} else {
-				fmt.Fprintf(w, "Pong failed\n")
-			}
-			return nil
-		}),
-	},
-}
-
-func pingPeer(ctx context.Context, n *core.IpfsNode, pid peer.ID, numPings int) <-chan interface{} {
-	outChan := make(chan interface{})
-	go func() {
-		defer close(outChan)
-
 		if len(n.Peerstore.Addrs(pid)) == 0 {
 			// Make sure we can find the node in question
-			select {
-			case outChan <- &PingResult{
+			if err := res.Emit(&PingResult{
 				Text:    fmt.Sprintf("Looking up peer %s", pid.Pretty()),
 				Success: true,
-			}:
-			case <-ctx.Done():
-				return
+			}); err != nil {
+				return err
 			}
 
-			ctx, cancel := context.WithTimeout(ctx, kPingTimeout)
-			defer cancel()
+			ctx, cancel := context.WithTimeout(req.Context, kPingTimeout)
 			p, err := n.Routing.FindPeer(ctx, pid)
+			cancel()
 			if err != nil {
-				select {
-				case outChan <- &PingResult{Text: fmt.Sprintf("Peer lookup error: %s", err)}:
-				case <-ctx.Done():
```

Comment:
This should have used the *command* context, not the find-peer context. This context is pretty much guaranteed to be expired at this point (so will return "success" to the user when we fail to lookup the peer). This is causing the JS interop tests to fail (for good reason).